knnvector/runners.py

```diff
@@ -0,0 +1,67 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+
+import logging
+
+from opensearchpy.exceptions import ConnectionTimeout
+
+from knnvector.tools.parse import parse_int_parameter, parse_string_parameter
+
+
+def register(registry):
+    registry.register_runner(
+        "custom-vector-bulk", BulkVectorsFromDataSetRunner(), async_runner=True
+    )
+    registry.register_runner(
+        "custom-refresh", CustomRefreshRunner(), async_runner=True
+    )
+
+
+class BulkVectorsFromDataSetRunner:
+
+    async def __call__(self, opensearch, params):
+        size = parse_int_parameter("size", params)
+        retries = parse_int_parameter("retries", params, 0) + 1
+
+        for _ in range(retries):
+            try:
+                await opensearch.bulk(
+                    body=params["body"]
+                )
+
+                return size, "docs"
+            except ConnectionTimeout:
+                logging.getLogger(__name__) \
+                    .warning("Bulk vector ingestion timed out. Retrying")
+
+        raise TimeoutError("Failed to submit bulk request in specified number "
+                           "of retries: {}".format(retries))
```

Comment:
why do we need this custom operation for ingestion? can't we use the Bulk ingestion operation provided in OSB?

Reply:
This was introduced to make sure that no documents are left out due to connection issue. This was observed during our testing. Since, recall assumes all documents are indexed completely, we added it originally.  I will add TODO to address this issue.